File: internal/account/route_authorization.go
package account

import (
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"fmt"
	"regexp"
	"strings"
	"time"

	"wedecent.com/wedecent/internal/mesh"
)

const maxRouteAuthorizationCost uint64 = 1_000_000_000

var routeAuthorizationDeviceIDPattern = regexp.MustCompile(`^wd_[a-z2-7]{16}$`)

type RouteAuthorizationRequest struct {
	SourceDeviceID      string
	RouterDeviceID      string
	DestinationDeviceID string
	FirstTransport      mesh.TransportName
	SecondTransport     mesh.TransportName
	FirstCost           uint64
	SecondCost          uint64
}

type routeAuthorizationResponse struct {
	Route         mesh.Route              `json:"route"`
	Authorization mesh.RouteAuthorization `json:"authorization"`
}

func (c Client) IssueRouteAuthorization(
	ctx context.Context,
	session *Session,
	request RouteAuthorizationRequest,
) (
	mesh.Route,
	mesh.RouteAuthorization,
	error,
) {
	if session == nil {
		return mesh.Route{},
			mesh.RouteAuthorization{},
			errors.New("account session is nil")
	}

	if err := session.validate(); err != nil {
		return mesh.Route{},
			mesh.RouteAuthorization{},
			fmt.Errorf(
				"invalid account session: %w",
				err,
			)
	}

	request = normalizeRouteAuthorizationRequest(request)

	if err := validateRouteAuthorizationRequest(request); err != nil {
		return mesh.Route{},
			mesh.RouteAuthorization{},
			err
	}

	var response routeAuthorizationResponse

	if err := c.doJSON(
		ctx,
		"POST",
		session.SupabaseURL+
			"/functions/v1/route-authorization",
		session.PublishableKey,
		session.AccessToken,
		map[string]any{
			"source_device_id":      request.SourceDeviceID,
			"router_device_id":      request.RouterDeviceID,
			"destination_device_id": request.DestinationDeviceID,
			"first_transport":       string(request.FirstTransport),
			"second_transport":      string(request.SecondTransport),
			"first_cost":            request.FirstCost,
			"second_cost":           request.SecondCost,
		},
		&response,
	); err != nil {
		return mesh.Route{},
			mesh.RouteAuthorization{},
			fmt.Errorf(
				"issue route authorization: %w",
				err,
			)
	}

	if err := validateRouteAuthorizationResponse(
		c.now().UTC(),
		request,
		response,
	); err != nil {
		return mesh.Route{},
			mesh.RouteAuthorization{},
			err
	}

	return response.Route,
		response.Authorization,
		nil
}

func normalizeRouteAuthorizationRequest(
	request RouteAuthorizationRequest,
) RouteAuthorizationRequest {
	request.SourceDeviceID =
		strings.TrimSpace(request.SourceDeviceID)

	request.RouterDeviceID =
		strings.TrimSpace(request.RouterDeviceID)

	request.DestinationDeviceID =
		strings.TrimSpace(request.DestinationDeviceID)

	request.FirstTransport =
		mesh.TransportName(
			strings.ToLower(
				strings.TrimSpace(
					string(request.FirstTransport),
				),
			),
		)

	request.SecondTransport =
		mesh.TransportName(
			strings.ToLower(
				strings.TrimSpace(
					string(request.SecondTransport),
				),
			),
		)

	return request
}

func validateRouteAuthorizationRequest(
	request RouteAuthorizationRequest,
) error {
	for name, id := range map[string]string{
		"source":      request.SourceDeviceID,
		"router":      request.RouterDeviceID,
		"destination": request.DestinationDeviceID,
	} {
		if !routeAuthorizationDeviceIDPattern.MatchString(id) {
			return fmt.Errorf(
				"invalid %s device ID",
				name,
			)
		}
	}

	if request.SourceDeviceID == request.RouterDeviceID ||
		request.SourceDeviceID == request.DestinationDeviceID ||
		request.RouterDeviceID == request.DestinationDeviceID {
		return errors.New(
			"source, router, and destination device IDs must differ",
		)
	}

	if !validRouteAuthorizationTransport(
		request.FirstTransport,
	) {
		return errors.New(
			"first routing transport must be lan or internet",
		)
	}

	if !validRouteAuthorizationTransport(
		request.SecondTransport,
	) {
		return errors.New(
			"second routing transport must be lan or internet",
		)
	}

	if request.FirstCost > maxRouteAuthorizationCost ||
		request.SecondCost > maxRouteAuthorizationCost {
		return errors.New(
			"route cost exceeds control-plane limit",
		)
	}

	return nil
}

func validRouteAuthorizationTransport(
	transport mesh.TransportName,
) bool {
	return transport == mesh.TransportLAN ||
		transport == mesh.TransportInternet
}

func validateRouteAuthorizationResponse(
	now time.Time,
	request RouteAuthorizationRequest,
	response routeAuthorizationResponse,
) error {
	route := response.Route

	if err := route.Validate(now); err != nil {
		return fmt.Errorf(
			"route authorization service returned invalid route: %w",
			err,
		)
	}

	if len(route.Hops) != 2 {
		return errors.New(
			"route authorization service returned a non-one-hop route",
		)
	}

	first := route.Hops[0]
	second := route.Hops[1]

	if route.Source !=
		mesh.DeviceID(request.SourceDeviceID) ||
		route.Destination !=
			mesh.DeviceID(request.DestinationDeviceID) ||
		first.From !=
			mesh.DeviceID(request.SourceDeviceID) ||
		first.To !=
			mesh.DeviceID(request.RouterDeviceID) ||
		first.Transport != request.FirstTransport ||
		first.Cost != request.FirstCost ||
		second.From !=
			mesh.DeviceID(request.RouterDeviceID) ||
		second.To !=
			mesh.DeviceID(request.DestinationDeviceID) ||
		second.Transport != request.SecondTransport ||
		second.Cost != request.SecondCost {
		return errors.New(
			"route authorization service returned a mismatched route",
		)
	}

	authorization := response.Authorization

	if strings.TrimSpace(authorization.KeyID) == "" ||
		len(authorization.KeyID) > 128 ||
		strings.ContainsAny(
			authorization.KeyID,
			"\r\n\t",
		) {
		return errors.New(
			"route authorization service returned an invalid key ID",
		)
	}

	signature, err :=
		base64.RawURLEncoding.DecodeString(
			authorization.Signature,
		)

	if err != nil ||
		len(signature) != ed25519.SignatureSize {
		return errors.New(
			"route authorization service returned an invalid signature",
		)
	}

	claims := authorization.Claims

	if claims.Version !=
		mesh.RouteAuthorizationVersion ||
		claims.Issuer !=
			mesh.RouteAuthorizationIssuer ||
		claims.Audience !=
			mesh.RouteAuthorizationAudience ||
		claims.Permission !=
			mesh.RouteAuthorizationPermission {
		return errors.New(
			"route authorization service returned an invalid capability profile",
		)
	}

	jti, err :=
		base64.RawURLEncoding.DecodeString(
			claims.JTI,
		)

	if err != nil || len(jti) != 16 {
		return errors.New(
			"route authorization service returned an invalid JTI",
		)
	}

	if claims.Router !=
		mesh.DeviceID(request.RouterDeviceID) {
		return errors.New(
			"route authorization service returned a mismatched router",
		)
	}

	if claims.IssuedAtUnixMS <= 0 ||
		claims.ExpiresUnixMS <=
			claims.IssuedAtUnixMS ||
		claims.ExpiresUnixMS-
			claims.IssuedAtUnixMS >
			mesh.MaxRouteAuthorizationLifetime.Milliseconds() ||
		claims.ExpiresUnixMS <= now.UnixMilli() {
		return errors.New(
			"route authorization service returned an invalid lifetime",
		)
	}

	if route.ExpiresAt.UTC().UnixMilli() !=
		claims.ExpiresUnixMS {
		return errors.New(
			"route authorization service returned mismatched expiry",
		)
	}

	if !routeAuthorizationRoutesEqual(
		route,
		claims.Route,
	) {
		return errors.New(
			"route authorization service returned mismatched signed route claims",
		)
	}

	return nil
}

func routeAuthorizationRoutesEqual(
	left mesh.Route,
	right mesh.Route,
) bool {
	if left.ID != right.ID ||
		left.Source != right.Source ||
		left.Destination != right.Destination ||
		left.ExpiresAt.UTC().UnixMilli() !=
			right.ExpiresAt.UTC().UnixMilli() ||
		len(left.Hops) != len(right.Hops) {
		return false
	}

	for index := range left.Hops {
		a := left.Hops[index]
		b := right.Hops[index]

		if a.From != b.From ||
			a.To != b.To ||
			a.Transport != b.Transport ||
			a.Cost != b.Cost {
			return false
		}
	}

	return true
}

File: supabase/functions/_shared/route-authorization.mjs
export const ROUTE_AUTHORIZATION_VERSION = 1
export const ROUTE_AUTHORIZATION_ISSUER = 'wedecent-control-plane'
export const ROUTE_AUTHORIZATION_AUDIENCE = 'wedecent-mesh-router'
export const ROUTE_AUTHORIZATION_PERMISSION = 'mesh.forward'
export const ROUTE_AUTHORIZATION_DOMAIN = 'wedecent-mesh-forward-v1'
export const MAX_ROUTE_AUTHORIZATION_LIFETIME_MS = 120_000
export const MAX_ROUTE_AUTHORIZATION_COST = 1_000_000_000

const deviceIDPattern = /^wd_[a-z2-7]{16}$/
const jtiPattern = /^[A-Za-z0-9_-]{22}$/
const encoder = new TextEncoder()

function fail(message) {
  throw new Error(`invalid route authorization: ${message}`)
}

function stringValue(value, name) {
  const text = String(value ?? '').trim()
  if (!text) fail(`${name} is required`)
  return text
}

function deviceID(value, name) {
  const text = stringValue(value, name)
  if (!deviceIDPattern.test(text)) fail(`${name} is invalid`)
  return text
}

function transport(value, name) {
  const text = stringValue(value, name).toLowerCase()
  if (text !== 'lan' && text !== 'internet') {
    fail(`${name} must be lan or internet`)
  }
  return text
}

function safeInteger(value, name, maximum = Number.MAX_SAFE_INTEGER) {
  const number = Number(value)
  if (
    !Number.isSafeInteger(number) ||
    number < 0 ||
    number > maximum
  ) {
    fail(`${name} is invalid`)
  }
  return number
}

function timestampMS(value, name) {
  const parsed = Date.parse(String(value ?? ''))
  if (!Number.isSafeInteger(parsed) || parsed <= 0) {
    fail(`${name} is invalid`)
  }
  return parsed
}

function encodedString(value) {
  const bytes = encoder.encode(value)
  if (bytes.length > 0xffffffff) fail('string field is too large')

  const framed = new Uint8Array(4 + bytes.length)
  new DataView(framed.buffer).setUint32(0, bytes.length, false)
  framed.set(bytes, 4)
  return framed
}

function encodedUint32(value) {
  const bytes = new Uint8Array(4)
  new DataView(bytes.buffer).setUint32(0, value, false)
  return bytes
}

function encodedUint64(value) {
  const big = BigInt(value)
  if (big < 0n || big > 0xffffffffffffffffn) {
    fail('uint64 field is out of range')
  }

  const bytes = new Uint8Array(8)
  new DataView(bytes.buffer).setBigUint64(0, big, false)
  return bytes
}

function concatenate(parts) {
  let length = 0
  for (const part of parts) length += part.length

  const output = new Uint8Array(length)
  let offset = 0

  for (const part of parts) {
    output.set(part, offset)
    offset += part.length
  }

  return output
}

export function encodeBase64URL(bytes) {
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)

  return btoa(binary)
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/g, '')
}

export function generateRouteAuthorizationJTI() {
  const raw = new Uint8Array(16)
  crypto.getRandomValues(raw)
  return encodeBase64URL(raw)
}

export function buildRouteAuthorizationClaims(row) {
  if (!row || typeof row !== 'object' || Array.isArray(row)) {
    fail('grant row is missing')
  }

  const id = stringValue(row.id, 'route id')
  const jti = stringValue(row.jti, 'jti')
  if (!jtiPattern.test(jti)) fail('jti is invalid')

  const source = deviceID(row.source_device_id, 'source device id')
  const router = deviceID(row.router_device_id, 'router device id')
  const destination = deviceID(
    row.destination_device_id,
    'destination device id'
  )

  if (
    source === router ||
    source === destination ||
    router === destination
  ) {
    fail('route devices must be distinct')
  }

  const firstTransport = transport(
    row.first_transport,
    'first transport'
  )
  const secondTransport = transport(
    row.second_transport,
    'second transport'
  )

  const firstCost = safeInteger(
    row.first_cost,
    'first cost',
    MAX_ROUTE_AUTHORIZATION_COST
  )
  const secondCost = safeInteger(
    row.second_cost,
    'second cost',
    MAX_ROUTE_AUTHORIZATION_COST
  )

  if (String(row.permission ?? '') !== ROUTE_AUTHORIZATION_PERMISSION) {
    fail('permission is invalid')
  }

  const issuedAtMS = timestampMS(row.issued_at, 'issued_at')
  const expiresAtMS = timestampMS(row.expires_at, 'expires_at')

  if (
    expiresAtMS <= issuedAtMS ||
    expiresAtMS - issuedAtMS > MAX_ROUTE_AUTHORIZATION_LIFETIME_MS
  ) {
    fail('lifetime is invalid')
  }

  const route = {
    ID: id,
    Source: source,
    Destination: destination,
    Hops: [
      {
        From: source,
        To: router,
        Transport: firstTransport,
        Cost: firstCost,
      },
      {
        From: router,
        To: destination,
        Transport: secondTransport,
        Cost: secondCost,
      },
    ],
    ExpiresAt: new Date(expiresAtMS).toISOString(),
  }

  return {
    route,
    claims: {
      v: ROUTE_AUTHORIZATION_VERSION,
      iss: ROUTE_AUTHORIZATION_ISSUER,
      aud: ROUTE_AUTHORIZATION_AUDIENCE,
      permission: ROUTE_AUTHORIZATION_PERMISSION,
      jti,
      router,
      route,
      iat_unix_ms: issuedAtMS,
      exp_unix_ms: expiresAtMS,
    },
  }
}

export function canonicalRouteAuthorizationMessage(input) {
  if (!input || typeof input !== 'object') {
    fail('authorization is missing')
  }

  const kid = stringValue(input.kid, 'key id')
  if (
    kid.length > 128 ||
    /[\r\n\t]/.test(kid)
  ) {
    fail('key id is invalid')
  }

  const claims = input.claims
  if (!claims || typeof claims !== 'object') {
    fail('claims are missing')
  }

  if (claims.v !== ROUTE_AUTHORIZATION_VERSION) {
    fail('version is invalid')
  }
  if (claims.iss !== ROUTE_AUTHORIZATION_ISSUER) {
    fail('issuer is invalid')
  }
  if (claims.aud !== ROUTE_AUTHORIZATION_AUDIENCE) {
    fail('audience is invalid')
  }
  if (claims.permission !== ROUTE_AUTHORIZATION_PERMISSION) {
    fail('permission is invalid')
  }

  const jti = stringValue(claims.jti, 'jti')
  if (!jtiPattern.test(jti)) fail('jti is invalid')

  const router = deviceID(claims.router, 'router')
  const route = claims.route

  if (!route || typeof route !== 'object') {
    fail('route is missing')
  }

  const routeID = stringValue(route.ID, 'route id')
  const source = deviceID(route.Source, 'route source')
  const destination = deviceID(
    route.Destination,
    'route destination'
  )

  if (!Array.isArray(route.Hops) || route.Hops.length !== 2) {
    fail('route must contain exactly two hops')
  }

  const issuedAtMS = safeInteger(
    claims.iat_unix_ms,
    'issued_at'
  )
  const expiresAtMS = safeInteger(
    claims.exp_unix_ms,
    'expires_at'
  )

  if (
    issuedAtMS <= 0 ||
    expiresAtMS <= issuedAtMS ||
    expiresAtMS - issuedAtMS > MAX_ROUTE_AUTHORIZATION_LIFETIME_MS
  ) {
    fail('lifetime is invalid')
  }

  const routeExpiryMS = Date.parse(String(route.ExpiresAt ?? ''))
  if (
    !Number.isSafeInteger(routeExpiryMS) ||
    routeExpiryMS !== expiresAtMS
  ) {
    fail('route expiry is not canonical')
  }

  const hops = route.Hops.map((hop, index) => ({
    From: deviceID(hop.From, `hop ${index} from`),
    To: deviceID(hop.To, `hop ${index} to`),
    Transport: transport(
      hop.Transport,
      `hop ${index} transport`
    ),
    Cost: safeInteger(
      hop.Cost,
      `hop ${index} cost`,
      MAX_ROUTE_AUTHORIZATION_COST
    ),
  }))

  if (
    hops[0].From !== source ||
    hops[0].To !== router ||
    hops[1].From !== router ||
    hops[1].To !== destination
  ) {
    fail('route/router binding is invalid')
  }

  const parts = [
    encodedString(ROUTE_AUTHORIZATION_DOMAIN),
    encodedUint32(ROUTE_AUTHORIZATION_VERSION),
  ]

  for (const value of [
    kid,
    claims.iss,
    claims.aud,
    claims.permission,
    jti,
    router,
    routeID,
    source,
    destination,
  ]) {
    parts.push(encodedString(value))
  }

  parts.push(
    encodedUint64(issuedAtMS),
    encodedUint64(expiresAtMS),
    encodedUint64(routeExpiryMS),
    encodedUint32(hops.length)
  )

  for (const hop of hops) {
    parts.push(
      encodedString(hop.From),
      encodedString(hop.To),
      encodedString(hop.Transport),
      encodedUint64(hop.Cost)
    )
  }

  return concatenate(parts)
}

export async function signRouteAuthorization(key, kid, row) {
  const built = buildRouteAuthorizationClaims(row)

  const message = canonicalRouteAuthorizationMessage({
    kid,
    claims: built.claims,
  })

  const rawSignature = new Uint8Array(
    await crypto.subtle.sign(
      'Ed25519',
      key,
      message
    )
  )

  if (rawSignature.length !== 64) {
    throw new Error('route authorization signer returned an invalid signature')
  }

  return {
    route: built.route,
    authorization: {
      kid,
      claims: built.claims,
      signature: encodeBase64URL(rawSignature),
    },
  }
}
